hypothesis-python/src/hypothesis/internal/observability.py

```diff
@@ -51,21 +47,68 @@
 from hypothesis.internal.escalation import InterestingOrigin
 from hypothesis.internal.floats import float_to_int
 from hypothesis.internal.intervalsets import IntervalSet
+from hypothesis.internal.validation import try_convert
 
 if TYPE_CHECKING:
     from hypothesis.internal.conjecture.data import ConjectureData, Spans, Status
 
-
 Observation: TypeAlias = Union["InfoObservation", "TestCaseObservation"]
 CallbackThreadT: TypeAlias = Callable[[Observation], None]
 # for all_threads=True, we pass the thread id as well.
 CallbackAllThreadsT: TypeAlias = Callable[[Observation, int], None]
 CallbackT: TypeAlias = CallbackThreadT | CallbackAllThreadsT
 
-# thread_id: list[callback]
-_callbacks: dict[int | None, list[CallbackThreadT]] = {}
-# callbacks where all_threads=True was set
-_callbacks_all_threads: list[CallbackAllThreadsT] = []
+
+_WROTE_TO = set()
+_deliver_to_file_lock = Lock()
+
+
+def _deliver_to_file(observation: Observation) -> None:  # pragma: no cover
+    from hypothesis.strategies._internal.utils import to_jsonable
+
+    kind = "testcases" if observation.type == "test_case" else "info"
+    fname = storage_directory("observed", f"{date.today().isoformat()}_{kind}.jsonl")
+    fname.parent.mkdir(exist_ok=True, parents=True)
+
+    observation_bytes = (
+        json.dumps(to_jsonable(observation, avoid_realization=False)) + "\n"
+    )
+    # only allow one conccurent file write to avoid write races. This is likely to make
+    # HYPOTHESIS_EXPERIMENTAL_OBSERVABILITY quite slow under threading. A queue
+    # would be an improvement, but that requires a background thread, and I
+    # would prefer to avoid a thread in the single-threaded case. We could
+    # switch over to a queue if we detect multithreading, but it's tricky to get
+    # right.
+    with _deliver_to_file_lock:
+        _WROTE_TO.add(fname)
+        with fname.open(mode="a") as f:
+            f.write(observation_bytes)
+
+
+@dataclass(slots=True, frozen=False)
+class ObservabilitySettings:
+    """
+    Options for the |settings.observability| argument to |@settings|.
+
+    Parameters
+    ----------
+
+    coverage : bool
+        Include the ``coverage`` field in test case observations.
+    choices : bool
+        Include the ``metadata.choice_nodes`` and ``metadata.choice_spans``
+        fields in test case observations.
+    callbacks: Collection[Callbable]
+        Observability callbacks. Each callback will be called for each observation
+        this test produces. See :ref:`observability <observability>` for details.
+    """
+
+    coverage: bool = True
+    choices: bool = False
+    callbacks: tuple[Callable, ...] = field(default=(_deliver_to_file,))
```

Comment:
```suggestion
    coverage: Literal["line", "branch"] | Collection[Literal["line", "branch"]] | None = None
    choices: Literal["values", "nodes"] | None = None
    write_file: bool = True
    callbacks: tuple[Callable[[dict], None], ...] = ()
```
plus validation to ensure that we always have `write_file or callbacks`, a `__or__` method to merge them, etc.

Alternatively we could make it easy to get the standard callback by making it a `staticmethod` on the `ObservabilitySettings` class 🤔

Reply:
> plus validation to ensure that callbacks is nonempty

done

I'm going to leave unioning and a staticmethod to a followup I think